contents/javascript/Javascript-framework-Redux.md

```diff
@@ -0,0 +1,228 @@
+{ type: 'SET_VISIBILITY_FILTER', filter: 'SHOW_ALL' }
+```
+
+Actions can be dispatched from user interactions in the application, such as button click events. For example, in our to-do list application, an `ADD_TODO` action could be dispatched when our the user fills up a task 'Go to swimming pool' and clicks an 'Add To-do' button.
+
+The current state, and an action is passed into a **reducer**, a function which returns the next state of the application
+
+Reducers are <tooltip content="A pure function is a function which: Given the same input, will always return the same output. And produces no side effects.">pure functions</tooltip>. They take the previous state and an action, and return the next state as new state objects, instead of mutating the previous state.
+
+With the updated application state, the application will reflect the new information in the store accordingly. For example, it may render new tasks, or render incomplete tasks instead of completed tasks.
+
+In a real-world application with a large and complex application state, having a single reducer to manage multiple parts of the state may get confusing and complicated. We can have multiple, simpler reducers instead, and combine them into one reducer.
+
+For example, we have two simpler reducers.
+
+Firstly, a reducer for the visibility filter `visibilityFilter`:
+
+```js
+function visibilityFilter(state = 'SHOW_ALL', action) {
+  if (action.type === 'SET_VISIBILITY_FILTER') {
+    return action.filter
+  } else {
+    return state
+  }
+}
+```
+
+And secondly, a reducer, that manages the todo-list items `todos`:
+
+```js
+function todos(state = [], action) {
+  switch (action.type) {
+    case 'ADD_TODO':
+      return state.concat([{ text: action.text, completed: false }])
+    case 'TOGGLE_TODO':
+      return state.map((todo, index) =>
+        action.index === index
+          ? { text: todo.text, completed: !todo.completed }
+          : todo
+      )
+    default:
+      return state
+  }
+}
+```
+
+And these two reducers are called by a main reducer. With that the complete application state is managed.
+
+```js
+function todoApp(state = {}, action) {
+  return {
+    todos: todos(state.todos, action),
+    visibilityFilter: visibilityFilter(state.visibilityFilter, action)
+  }
+}
+```
+
+A careful eye would notice that the above code block examples on reducers are not pure functions. If you're interested, you may see how the reducer should actually be written in the [official Redux site](https://redux.js.org/basics/reducers/#source-code). The code in the link includes some Redux API, and Javascript methods such as [`.map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) and the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) to make the functions pure. But knowing the actual syntax for Redux is not the main focus of this introductory article.
```

Comment:
Feels like reducers are claimed to be 'pure functions' in some places and also 'not pure functions' in other places.